chain-abstraction/sol-implementation/edge_challenge_manager.go

```diff
@@ -328,62 +327,44 @@ func (e *specEdge) ConfirmByClaim(ctx context.Context, claimId protocol.ClaimId)
-		}
-		if blockChallengeOneStepForkSource.IsNone() {
-			return protocol.OriginHeights{}, errors.New("source edge is none")
-		}
-		startHeight, _ := blockChallengeOneStepForkSource.Unwrap().StartCommitment()
+	challengeLevel, err := e.GetChallengeLevel()
+	if err != nil {
+		return protocol.OriginHeights{}, err
+	}
+	if challengeLevel == 0 {
+		startHeight, _ := e.StartCommitment()
 		return protocol.OriginHeights{
-			BlockChallengeOriginHeight: startHeight,
+			ChallengeOriginHeights: []protocol.Height{startHeight},
 		}, nil
-	case protocol.SmallStepChallengeEdge:
-		rivalId, err := e.manager.caller.FirstRival(&bind.CallOpts{Context: ctx}, e.inner.OriginId)
+	}
+	challengeOriginHeights := make([]protocol.Height, challengeLevel)
+	originId := e.inner.OriginId
+	for challengeLevel > 0 {
+		rivalId, err := e.manager.caller.FirstRival(&bind.CallOpts{Context: ctx}, originId)
 		if err != nil {
 			return protocol.OriginHeights{}, err
 		}
-		bigStepChallengeOneStepForkSource, err := e.manager.GetEdge(ctx, protocol.EdgeId{Hash: rivalId})
+		challengeOneStepForkSource, err := e.manager.GetEdge(ctx, protocol.EdgeId{Hash: rivalId})
 		if err != nil {
-			return protocol.OriginHeights{}, errors.Wrap(err, "big step challenge one step fork source does not exist")
+			return protocol.OriginHeights{}, errors.Wrapf(err, "big step challenge one step fork source does not exist: origin id %#x, rival %#x, challenge level %d", originId, rivalId, challengeLevel)
 		}
-		if bigStepChallengeOneStepForkSource.IsNone() {
+		if challengeOneStepForkSource.IsNone() {
 			return protocol.OriginHeights{}, errors.New("source edge is none")
 		}
-		bigStepEdge, ok := bigStepChallengeOneStepForkSource.Unwrap().(*specEdge)
+		bigStepEdge, ok := challengeOneStepForkSource.Unwrap().(*specEdge)
 		if !ok {
 			return protocol.OriginHeights{}, errors.New("not *SpecEdge")
 		}
-		rivalId, err = e.manager.caller.FirstRival(&bind.CallOpts{Context: ctx}, bigStepEdge.inner.OriginId)
-		if err != nil {
-			return protocol.OriginHeights{}, err
-		}
-		blockChallengeOneStepForkSource, err := e.manager.GetEdge(ctx, protocol.EdgeId{Hash: rivalId})
-		if err != nil {
-			return protocol.OriginHeights{}, errors.Wrap(err, "block challenge one step fork source does not exist")
-		}
-		if blockChallengeOneStepForkSource.IsNone() {
-			return protocol.OriginHeights{}, errors.New("source edge is none")
-		}
 		bigStepStartHeight, _ := bigStepEdge.StartCommitment()
-		blockChallengeStartHeight, _ := blockChallengeOneStepForkSource.Unwrap().StartCommitment()
-		return protocol.OriginHeights{
-			BlockChallengeOriginHeight:   blockChallengeStartHeight,
-			BigStepChallengeOriginHeight: bigStepStartHeight,
-		}, nil
-	default:
-		startHeight, _ := e.StartCommitment()
-		return protocol.OriginHeights{
-			BlockChallengeOriginHeight: startHeight,
-		}, nil
+
+		challengeOriginHeights[challengeLevel-1] = bigStepStartHeight
+		originId = bigStepEdge.inner.OriginId
```

Comment:
any concern that `inner` could be nil?

Reply:
it isn't a pointer, so we'll be fine here